pkg/api/validation/validation.go

```diff
@@ -1346,6 +1350,87 @@ func ValidatePodSpec(spec *api.PodSpec, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
+// ValidateNodeSelectorRequirement test that the specified NodeSelectorRequirement fields has valid data
+func ValidateNodeSelectorRequirement(rq api.NodeSelectorRequirement, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	switch rq.Operator {
+	case api.NodeSelectorOpIn, api.NodeSelectorOpNotIn:
+		if len(rq.Values) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("values"), "must be specified when `operator` is 'In' or 'NotIn'"))
+		}
+	case api.NodeSelectorOpExists, api.NodeSelectorOpDoesNotExist:
+		if len(rq.Values) > 0 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("values"), "may not be specified when `operator` is 'Exists' or 'DoesNotExist'"))
+		}
+
+	case api.NodeSelectorOpGt, api.NodeSelectorOpLt:
+		if len(rq.Values) != 1 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("values"), "must be specified single value when `operator` is 'Lt' or 'Gt'"))
+		}
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("operator"), rq.Operator, "not a valid selector operator"))
+	}
+	allErrs = append(allErrs, ValidateLabelName(rq.Key, fldPath.Child("key"))...)
+	return allErrs
+}
+
+// ValidateNodeSelector test that the specified nodeSelector fields has valid data
+func ValidateNodeSelector(nodeSelector *api.NodeSelector, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for _, nst := range nodeSelector.NodeSelectorTerms {
```

Comment:
Now that we have validation, I think maybe we should reject
len(NodeSelector.NodeSelectorTerms) == 0
and also reject
len(NodeSelectorTerm.MatchExpressions) == 0
and also reject
len(NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) == 0
and also reject
len(PreferredSchedulingTerm.MatchExpressions) == 0

I think those situations would only happen when user makes an error in their configuration, so it would be good to return error to the user. I realize we have rules on how to handle those scenarios right now, and I think it's OK to leave those rules in the code (i.e. I would not change any of the code you have in scheduler predicate or priority functions), but we can return error during validation (so the lower levels of the code should never see those situations). 

If you do that, then I would also remove (from types.go) the comments "A null or empty node selector term matches no objects" and "An empty preferred scheduling term matches all objects with implicit weight 0 (i.e. it's a no-op). A null preferred scheduling term matches no objects (i.e. is also a no-op)."

Does that sound reasonable?